### PR TITLE
Report Form (With Library)

### PR DESCRIPTION
**Pull request by justinkonecny**
## Why

[Monday.com Ticket](https://code4community-team.monday.com/boards/866744728/pulses/1058514653)

Create "Report With Library" page for submitting report forms.

## This PR

- Update `FormFooter`
  - Parent component now supplies previous route and next route (optional)
  - Update "Submit" / "Next" to be one button or the other
- Update navigation bar space and size
- Update `LibraryInfo` container to be its own page
  - Redirects to `ReportWithLibrary` container if "Yes"
- Add `ReportWithLibrary` container and component
- Fix `Row` spacing for all items (add vertical spacing between rows)

## Screenshots

Screen Shot 2021-03-30 at 10 09 59 PM

Screen Shot 2021-03-30 at 10 10 13 PM

Screen Shot 2021-03-30 at 10 10 26 PM

Screen Shot 2021-03-30 at 10 10 39 PM

Screen Shot 2021-03-30 at 10 10 59 PM

### CODE REVIEW

File: src/components/report/TrainingMentorshipInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import FormContainer from '../../components/form-style/FormContainer';
+import FormPieceBoolean from '../form-style/FormPieceBoolean';
+
+const TrainingMentorshipInfo: React.FC = () => {
+  const [isStudentLibrary, setIsStudentLibrary] = useState<boolean>();
+  const [involvedParents, setInvolvedParents] = useState<boolean>();
+  const [
+    teachersSeekingSupport,
+    setTeacherSeekingSupport,
+  ] = useState<boolean>();
```
Comment (jackblanc) [inline src/components/report/TrainingMentorshipInfo.tsx:13]: Why do we need useState here and not elsewhere? Where do we read these values?

File: src/containers/reportWithLibrary/ducks/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+} from './types';
+
+export const getReportWithLibrary = (
+  state: ReportWithLibraryReducerState,
+): ReportWithLibraryResponse | null => {
+  if (asyncRequestIsComplete(state.reportWithLibrary)) {
+    return state.reportWithLibrary.result;
+  }
+  return null;
+};
```
Comment (jackblanc) [inline src/containers/reportWithLibrary/ducks/selectors.ts:14]: This selector essentially just reduces clarity of data IMO. We go from one of four states to one of two, losing the context of why there is no form response. We should check the status in the containers.

File: src/containers/reportWithLibrary/ducks/types.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  readonly apprenticeshipProgram: null | string;
+  readonly trainsAndMentorsApprentices: null | boolean;
+  readonly hasCheckInTimetables: null | boolean;
+  readonly hasBookCheckoutSystem: null | boolean;
+  readonly numberOfStudentLibrarians: null | number;
+  readonly reasonNoStudentLibrarians: null | string;
+  readonly hasSufficientTraining: null | boolean;
+  readonly teacherSupport: null | string;
+  readonly parentSupport: null | string;
+  readonly visitReason: null | string;
```
Comment (jackblanc) [inline src/containers/reportWithLibrary/ducks/types.ts:35]: use `readonly field?: string` for optional types.

File: src/containers/reportWithLibrary/ducks/types.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+export interface ReportWithLibraryResponse extends ReportWithLibraryRequest {
+  readonly id: number;
+  readonly schoolId: number;
+  readonly libraryStatus: string;
+}
+
+export enum AssignedPersonRole {
```
Comment (jackblanc) [inline src/containers/reportWithLibrary/ducks/types.ts:44]: This feels like something that should be keyed to contacts.

File: src/containers/reportWithLibrary/index.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+const ReportWithLibraryContainer = () => {
+  const dispatch = useDispatch();
+  //
+  // TODO: Add functionality for loading existing data
+  //
+  //
+  // useEffect(() => {
+  //   dispatch(loadReportWithLibrary(reportWithLibraryId));
+  // }, [reportWithLibraryId]);
```
Comment (jackblanc) [inline src/containers/reportWithLibrary/index.tsx:15]: pretty much the right idea yea.

File: src/utils/helpers.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { Select } from 'antd';
+
+const { Option } = Select;
+
+export const toTitleCase = (str: string): string => {
```
Comment (jackblanc) [inline src/utils/helpers.tsx:6]: These need tests, as well as everything in ducks.

**Review by jackblanc — COMMENTED** [review]
This looks pretty good, it's hard to get a read on it because of the size of the PR. Let's try to keep changes under +/- 500 lines.

File: src/components/report/TrainingMentorshipInfo.tsx (hunk 1 above)
Comment (justinkonecny) [reply src/components/report/TrainingMentorshipInfo.tsx:13]: The original component had this logic for conditionally displaying fields to the user. Basically, we only show the "explain here" input fields if the original question is answered a certain way.

File: src/containers/reportWithLibrary/ducks/types.ts (hunk 3 above)
Comment (justinkonecny) [reply src/containers/reportWithLibrary/ducks/types.ts:35]: The fields themselves are required by the backend, but they can be `null`?

Comment (jackblanc) [reply src/containers/reportWithLibrary/ducks/types.ts:35]: Oh I see the difference

File: src/api/protectedApiClient.ts [hunk 7]
```diff
@@ -110,12 +124,39 @@ const deleteSchoolContact = (
     .catch((err) => err);
 };
 
+const getReportWithLibrary = (
+  reportId: number,
+): Promise<ReportWithLibraryResponse> => {
+  return AppAxiosInstance.get(
+    `${ProtectedApiClientRoutes.REPORT_WITH_LIBRARY}/${reportId.toString()}`,
```
Comment (wwohlers) [inline src/api/protectedApiClient.ts:131]: I don't see this route anywhere in the Reports API spec. https://docs.c4cneu.com/hats/api-spec-school-report/ Is this an actual route? It seems like we can either get all reports or the latest one, not a specific one by ID.

File: src/components/form-style/FormFooter.tsx [hunk 8]
```diff
@@ -3,84 +3,65 @@ import { Row, Col, Form, Button } from 'antd';
...
+              {props.areAbleToSubmit ? (
+                <Button type="primary" htmlType="submit" ghost>
+                  Submit
+                </Button>
+              ) : (
+                <LinkButton
+                  to={props.next || Routes.HOME}
+                  disabled={disableNext}
+                >
+                  Next Section <ArrowRightOutlined />
```
Comment (wwohlers) [inline src/components/form-style/FormFooter.tsx:58]: I believe we agreed to put the navigation buttons inside the containers, no? I could be mistaken, and it's not a big deal either way, but we definitely need to pick one and stick with it.

File: src/components/report/TrainingMentorshipInfo.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
+import React, { useState } from 'react';
+import { Col, Form, InputNumber, RadioChangeEvent, Row } from 'antd';
+import { ClarifyText, FormTextArea } from '../index';
+import FormContainer from '../../components/form-style/FormContainer';
+import FormPieceBoolean from '../form-style/FormPieceBoolean';
+
+const TrainingMentorshipInfo: React.FC = () => {
+  const [isStudentLibrary, setIsStudentLibrary] = useState<boolean>();
+  const [involvedParents, setInvolvedParents] = useState<boolean>();
```
Comment (wwohlers) [inline src/components/report/TrainingMentorshipInfo.tsx:9]: Should call this `isInvolvedParents` to follow naming convention

File: src/containers/reportWithLibrary/ducks/thunks.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+export const createReportWithLibrary = (
+  schoolId: number,
+  report: ReportWithLibraryRequest,
+): ReportWithLibraryThunkAction<void> => {
+  return (dispatch, getState, { protectedApiClient }) => {
+    return protectedApiClient
+      .createReportWithLibrary(schoolId, report)
+      .then(() => {
+        dispatch(loadReportWithLibrary(schoolId));
```
Comment (wwohlers) [inline src/containers/reportWithLibrary/ducks/thunks.ts:32]: You're passing in the schoolId as an argument to loadReportWithLibrary, but that thunks takes in reportId as its argument. I think the loadReportWithLibrary should take in a schoolId (and the getReportWithLibrary endpoint should also take in a schoolId).

**Review by wwohlers — CHANGES_REQUESTED** [review]
Looks good for the most part, a couple important things.

First, it seems like the methods you added to the protectedApiClient don't quite match with the API spec. The spec could be wrong, but something I noticed.

Another thing is I believe that we agreed to leave the responsibility of navigation to the containers instead of using a generic FormFooter. We should pick one way and stick with it. 

Last thing is more of an opinion/stylistic thing, but this PR introduces a lot of changes that aren't directly related to the report with library container. It's not a problem in and of itself but it does make it much harder to review. @jackblanc I'm interested what you think about this but I think in general in PRs we should only make changes that are directly relevant to the Monday ticket it addresses.

**Pull request closed without merging**